### PR TITLE
asserts: db support for exporting pub keys and making/signing assertions

### PR DESCRIPTION
**Pull request by pedronis**
implements Sign() and ExportPublicKey() on the Database, use fingerprints or suffixes thereof to find the key pair to use,

also introduce EncodePublicKey()

also finish consolidating things into crypto.go so that only crypto.go and tests use golang.org/x/openpgg/packet directly

first pass at changing tests to use ExportPublicKey(), EncodePublicKey() and Sign() where it makes sense, removes some fiddly bits from them as the removal of some of the stdlib imports in them shows

will add more tests about the unhappy paths of the new methods/functions in a subsequent branch

Also this branch shows ( between the lines ) that PublicKey is conflating too many features/roles, we probably need to split it at least in two facets/interfaces (for example PublicKey more like the crypto-only oriented Signature and PrivateKey, and SignatureVerifier) but probably after current demo

### CODE REVIEW

File: asserts/database.go [hunk 1]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
-	err = db.atomicWriteEntry(buf.Bytes(), true, privateKeysRoot, authorityID, fingerp)
+	fingerp := privKey.PublicKey().Fingerprint()
+	err = db.atomicWriteEntry(encoded, true, privateKeysRoot, authorityID, fingerp)
 	if err != nil {
 		return "", fmt.Errorf("failed to store private key: %v", err)
 	}
 	return fingerp, nil
 }
 
+// finding 0 or more than one private key are considered errors
```
Comment (mvo5) [inline asserts/database.go]: (nitpick) I had trouble parsing this line, maybe reword a little bit like "findPrivateKey will returns an error if none or more than one private keys are found"?

File: asserts/database.go [hunk 2]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+	foundPrivKeyCb := func(relpath string) error {
+		if keyPath != "" {
+			return fmt.Errorf("ambiguous search, more than one key pair found: %q and %q", keyPath, relpath)
+
+		}
+		keyPath = relpath
+		return nil
+	}
+	privKeysTop := filepath.Join(db.root, privateKeysRoot)
+	err := findWildcard(privKeysTop, []string{authorityID, fingerprintWildcard}, foundPrivKeyCb)
```
Comment (mvo5) [inline asserts/database.go]: Nice use of the callback and its scoping!

**Comment by mvo5** [comment]
This looks very good to me. However Gustavo should probably also have a look just to ensure its all aligned with him.

File: asserts/database.go (hunk 1 above)
Comment (chipaca) [inline asserts/database.go]: "If not exactly one is found"?

**Comment by chipaca** [comment]
LGTM. A little bit harder to review for me than I'd like, because there's a mini refactor mixed in with the features, but that's very minor. :+1:

File: asserts/crypto.go [hunk 3]
```diff
@@ -28,26 +28,77 @@ import (
...
+const (
+	maxEncodeLineLength = 76
+)
 
-func generatePrivateKey() (*packet.PrivateKey, error) {
-	priv, err := rsa.GenerateKey(rand.Reader, 2048)
+func encodeFormatAndData(format string, data []byte) []byte {
+	buf := bytes.NewBufferString(format)
+	buf.WriteByte(' ')
+	buf.Grow(base64.StdEncoding.EncodedLen(len(data)))
```
Comment (niemeyer) [inline asserts/crypto.go]: Somewhat of a nitpick, but if we're growing, why are we doing that after doing operations that must have just re-allocated a buffer internally?  `Grow(enclen(data)+1+len(format))`?

File: asserts/crypto.go [hunk 4]
```diff
@@ -178,3 +230,55 @@ func parsePublicKey(pubKey []byte) (PublicKey, error) {
...
+	}
+	privk, ok := pkt.(*packet.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("expected private key, got instead: %T", pkt)
+	}
+	return openpgpPrivateKey{privk}, nil
+}
+
+// WrapPrivateKey returns a PrivateKey for database use out of a opengpg packet.PrivateKey.
+func WrapPrivateKey(privk *packet.PrivateKey) PrivateKey {
```
Comment (niemeyer) [inline asserts/crypto.go]: s/Wrap/OpenPGP/?  Naming it "Wrap" makes it feel like PrivateKey is just a shell around the other, which might be technically right today, but it's not the goal of this abstraction. PrivateKey is our rock, and the openpgp variant is something we're choosing as the underlying implementation for the first version of it.

File: asserts/crypto.go [hunk 5]
```diff
@@ -28,26 +28,77 @@ import (
...
+
+	return buf.Bytes()
+}
+
+type openpgpSerializer interface {
+	openpgpSerialize(w io.Writer) error
+}
+
+func encodeOpenpgp(obj interface{}, kind string) ([]byte, error) {
+	serializer, ok := obj.(openpgpSerializer)
```
Comment (niemeyer) [inline asserts/crypto.go]: Why isn't obj itself typed as `openpgpSerializer`?

Comment (pedronis) [inline asserts/crypto.go]: @niemeyer because otherwise the check would need to be repeated in the callers, and we need the check until we split PublicKey along two different interfaces, which I think we need for clarity, but I want to do that after the demo is out

File: asserts/crypto.go [hunk 6]
```diff
@@ -28,26 +28,77 @@ import (
...
+	buf.Grow(flatSize + newlines)
+	off := 0
+	for i := 0; i < lines; i++ {
+		buf.Write(flat[off : off+maxEncodeLineLength])
+		buf.WriteByte('\n')
+		off += maxEncodeLineLength
+	}
+	if rest > 0 {
+		buf.Write(flat[off:])
+		buf.WriteByte('\n')
```
Comment (niemeyer) [inline asserts/crypto.go]: The logic above felt slightly too long/error prone for the goal. This would be equivalent to something along the lines of:
```
buf.Grow(len(flat) + len(flat)/maxEncodedLineLength + 1)
start := 0
for start < len(flat) {
        end := start + maxEncodeLineLength
        if end > len(flat) {
                end = len(flat)
        }
        buf.Write(flat[start:end])
        buf.WriteByte('\n')
        start = end
}

File: asserts/crypto.go (hunk 4 above)
Comment (niemeyer) [inline asserts/crypto.go]: (same for the public version previously exposed above)

File: asserts/crypto.go [hunk 7]
```diff
@@ -178,3 +230,55 @@ func parsePublicKey(pubKey []byte) (PublicKey, error) {
...
+
+type openpgpPrivateKey struct {
+	privk *packet.PrivateKey
+}
+
+func (opgPrivK openpgpPrivateKey) PublicKey() PublicKey {
+	return WrapPublicKey(&opgPrivK.privk.PublicKey)
+}
+
+func (opgPrivK openpgpPrivateKey) openpgpSerialize(w io.Writer) error {
```
Comment (niemeyer) [inline asserts/crypto.go]: This makes me wonder if what we want isn't actually `keySerialize(w io.Writer)` and `keyFormat() string`. This turns the `openpgpSerializer` interface into `keySerializer`, and `openpgpEncode` above into `keyEncode`, reducing the amount of "openpgp" we have in the code.

File: asserts/database.go [hunk 8]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+		return nil, err
+	}
+	if keyPath == "" {
+		return nil, fmt.Errorf("no matching key pair found")
+	}
+	encoded, err := db.readEntry(privateKeysRoot, keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read key pair: %v", err)
+	}
+	privKey, err := parsePrivateKey(encoded)
```
Comment (niemeyer) [inline asserts/database.go]: Should we `s/parse/decode/`, to make the two ends symmetrical?

File: asserts/crypto.go (hunk 7 above)
Comment (pedronis) [inline asserts/crypto.go]: @niemeyer we can do that, I thought about it, is slightly more code atm

File: asserts/database.go [hunk 9]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode key pair: %v", err)
+	}
+	return privKey, nil
+}
+
+// ExportPublicKey exports the public part of a stored key pair for identity
+// by matching the given fingerprint suffix, it is an error if no or more
+// than one key pair is found.
+func (db *Database) ExportPublicKey(authorityID string, fingerprintSuffix string) (PublicKey, error) {
```
Comment (niemeyer) [inline asserts/database.go]: `s/Export//`?

File: asserts/database.go [hunk 10]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+	privKey, err := db.findPrivateKey(authorityID, "*"+fingerprintSuffix)
```
Comment (niemeyer) [inline asserts/database.go]: Can `fingerprintSuffix` please be validated for being a proper suffix?  We shouldn't have to think about all the consequences of having something like a `*` reaching this code path.

File: asserts/database.go (hunk 9 above)
Comment (pedronis) [inline asserts/database.go]: @niemeyer again I thought of that, happier to move to that when PublicKey is a bit less overloaded

File: asserts/database.go [hunk 11]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+	if err != nil {
+		return nil, err
+	}
+	return privKey.PublicKey(), nil
+}
+
+// Sign makes an assertion with headers and body and signs it using the matching authority-id (from headers) key pair with the given fingeprint.
```
Comment (niemeyer) [inline asserts/database.go]: s/makes/builds/, to avoid the ambiguity of "making an assertion"

Comment (niemeyer) [inline asserts/database.go]: Also, typo on fingerprint

Comment (niemeyer) [inline asserts/database.go]: Sorry, I'll nitpick further here, since the sentence as a whole feels hard to read.. let me propose something:
```
// Sign builds an assertion with the provided information and signs it.
// The assertion is signed with the private key from `headers["authority-id"]` that has the provided fingerprint.

File: asserts/database.go [hunk 12]
```diff
@@ -125,26 +121,86 @@ func (db *Database) GenerateKey(authorityID string) (fingerprint string, err err
...
+// Fingerprint can be empty but then exactly one key pair
```
Comment (niemeyer) [inline asserts/database.go]: That feels like trouble coming. We'll end up building code that seems to work until a second key shows up.

**Comment by niemeyer** [comment]
Okay, quite a few comments, but the vast majority is trying to polish the interface. Only a couple are actual relevant issues which would be good to fix before landing this.

File: asserts/crypto.go (hunk 5 above)
Comment (niemeyer) [inline asserts/crypto.go]: We discussed how to solve that in our call (separate AccounKey from PublicKey, have encodeKey in both PublicKey and PrivateKey). Fine to merge without it being addressed for the time being.

File: asserts/crypto.go (hunk 7 above)
Comment (niemeyer) [inline asserts/crypto.go]: Agreed solution described in previous comment.

File: asserts/database.go (hunk 9 above)
Comment (niemeyer) [inline asserts/database.go]: Per our call, let's rename it now even if the untangling is a follow up.

File: asserts/asserts.go [hunk 13]
```diff
@@ -333,6 +330,10 @@ func buildAndSign(assertType AssertionType, headers map[string]string, body []by
 	if err != nil {
 		return nil, fmt.Errorf("failed to sign assertion: %v", err)
 	}
+	// add an extra newline
+	temp := bytes.NewBuffer(signature)
+	temp.WriteByte('\n')
+	signature = temp.Bytes()
```
Comment (niemeyer) [inline asserts/asserts.go]: signature = append(signature, '\n')

Comment (niemeyer) [inline asserts/asserts.go]: Btw, the comment says exactly what the code is saying. The more interesting question to answer in the comment is why the extra line is being added.

**Comment by niemeyer** [comment]
LGTM

File: asserts/asserts.go (hunk 13 above)
Comment (pedronis) [inline asserts/asserts.go]: fixed
